### PR TITLE
Apply config encoding at the ProviderWithContext level

### PR DESCRIPTION
**Pull request by iwahbe**
This allows dynamic providers to benefit from the same code that handles provider config for PF and SDKv2 based providers.

Fixes https://github.com/pulumi/pulumi-terraform-provider/issues/4

This borrows some code from <https://github.com/pulumi/pulumi-terraform-bridge/pull/2195>,
converted to still work with `plugin.ProviderWithContext`.

---

### Why the old design was problematic

The current design of provider encoding for the PF is:

``` go
type providerServer struct {
	pulumirpc.UnsafeResourceProviderServer // opt out of forward compat
	provider      ProviderWithContext
	keepSecrets   bool
	keepResources bool

	configEncoding tfbridge.ConfigEncoding
}
```
([src](https://github.com/pulumi/pulumi-terraform-bridge/blob/b40324a6a05cb68fbb61d29bdf2b7d7e5899e766/pf/internal/plugin/provider_server.go#L36-L44))

`*providerServer` directly implements `pulumirpc.ResourceProviderServer`. `providerServer`
is in charge of translating between `grpc` level constructs and `plugin.Provider` level
constructs. One method on `pulumirpc.ResourceProviderServer` is `Parameterize`:

``` go
type ResourceProviderServer interface {
	...
    Parameterize(context.Context, *ParameterizeRequest) (*ParameterizeResponse, error)
}
```
([src](https://github.com/pulumi/pulumi/blob/cfd74eaf031b7005f6d2b7887a3e6676e8b6a993/sdk/proto/go/provider_grpc.pb.go#L322))

This request is forwarded to `providerServer.provider`:

``` go
func (p *providerServer) Parameterize(
	ctx context.Context, req *pulumirpc.ParameterizeRequest,
) (*pulumirpc.ParameterizeResponse, error) {
	return plugin.NewProviderServer(&forwardServer{
		parameterize: p.provider.ParameterizeWithContext,
	}).Parameterize(ctx, req)
}
```
([src](https://github.com/pulumi/pulumi-terraform-bridge/blob/b40324a6a05cb68fbb61d29bdf2b7d7e5899e766/pf/internal/plugin/provider_server.go#L157-L163))

The problem is that `Parameterize` changes the set of acceptable configuration inputs, and
thus the encoding. With the old design, `Parameterize` would need a hook to change the
encoding used by `providerServer`.

### Why the new design

By moving `configEncoding` into `ProviderWithContext`, we allow `ProviderWithContext` to
change the encoding as calls to `Parameterize` come in. This is the change that fixes
<https://github.com/pulumi/pulumi-terraform-provider/issues/4>.

The new code also moves us in the direction of #2148, but #2148 does not motivate the
change.

#### [pulumi/pulumi: [sdk-gen] Don't JSON-encode provider config #15032](https://github.com/pulumi/pulumi/pull/15032)

This would be a better fix, however the core team has shown little apatite for merging
[#15032](https://github.com/pulumi/pulumi/pull/15032).  I don't want to wait on them to become interested and leave this unfixed
indefinitely.

### CODE REVIEW

File: pkg/tfbridge/config_encoding.go
Comment (t0yv0) [inline pkg/tfbridge/config_encoding.go:102]: These are the same changes that were making me uncomfortable in https://github.com/pulumi/pulumi-terraform-bridge/pull/2195 - can we please hold until I get a chance to review this step by step. It seems to me that it's fairly brittle part of the system and it's not been adequately tested.

**Comment by t0yv0** [comment]
I'd also suggest looking into @blampe work https://github.com/pulumi/pulumi/pull/15032 that's by far the cleaner fix. This gets rid of config encoding altogether. Even if we're unable to roll it out as default for providers, perhaps it could be an option in codegen and have dynamic providers use that option for local SDK generator?

**Comment by VenelinMartinov** [comment]
@iwahbe can we please update the PR description with an explanation of why we need to add an extra layer of nesting for the PF bridge to fix a dynamic provider issue?

Also is the config encoding tested by integration tests for PF/SDKv2? Seems like this affects both and I'd be wary of making changes to any of that without backfilling test coverage.

**Comment by t0yv0** [comment]
It's tested somewhat, on the happy path, but we could do more. Perhaps if we want to make a change to config encoding, we can do a quick sync review session? Or break down the changes step by step for an easier review? I'm open to making the change if https://github.com/pulumi/pulumi/pull/15032 is not landing in an expedient fashion.

**Comment by blampe** [comment]
> I'd also suggest looking into @blampe work [pulumi/pulumi#15032](https://github.com/pulumi/pulumi/pull/15032) that's by far the cleaner fix. This gets rid of the unnatural config encoding altogether. Even if we're unable to roll it out as default for providers, perhaps it could be an option in codegen and have dynamic providers use that option for local SDK generator?

Yeah, I'm pretty strongly opposed to adding any of the legacy config encoding stuff here because it will effectively mean that we must continue to support this weird v1 JSON encoding forever. Let's spend this time fixing the root problem.

**Comment by iwahbe** [comment]
> @iwahbe can we please update the PR description with an explanation of why we need to add an extra layer of nesting for the PF bridge to fix a dynamic provider issue?

Done

> Also is the config encoding tested by integration tests for PF/SDKv2? Seems like this affects both and I'd be wary of making changes to any of that without backfilling test coverage.

Both under test, but not in an extremely thorough way. Most of the testing lives in `config_encoding_test.go`.

**Comment by iwahbe** [comment]
> It's tested somewhat, on the happy path, but we could do more as I don't have full confidence. 

What additional tests would you like to see?

> There were some bw-compat behaviors in it specifically that I recall prevented turnaround from working. New code seems to assume turnaround. But maybe these are moot. 

Can you point to where the new code assumes turnaround? I don't think this change makes any new assumptions.

> Perhaps if we want to make a change to config encoding, we can do a quick sync review session? Or break down the changes step by step for an easier review? I'm open to making the change if [pulumi/pulumi#15032](https://github.com/pulumi/pulumi/pull/15032) is not landing in an expedient fashion.

I'll schedule a sync review session.

File: pf/internal/plugin/provider_server.go
Comment (iwahbe) [reply pf/internal/plugin/provider_server.go]: I have changed the comment to:

```
	// The bridge (and [providerServer] more specifically) attempts to respect the
	// hand-shake in performed in [Configure]. Unfortunately, [CheckConfig] happens
	// before [Configure] so we just assume that the engine supports secrets.
	//
	// TODO[pulumi/pulumi#16876]: This works around a protocol problem.
```

I have opened pulumi/pulumi#16876 to track the underlying protocol issue.

**Comment by iwahbe** [comment]
@t0yv0 I have split up the PR into 3 separate commits. 

Each commit description explains what the commit does.

File: pkg/tfbridge/config_encoding.go
Comment (t0yv0) [reply pkg/tfbridge/config_encoding.go:102]: Ok looks like this is a dead code, but it's concerning to lose this in refactoring. Let's maybe throw in an assert that opts.KeepSecrets is false here?

File: dynamic/provider_test.go
Comment (t0yv0) [inline dynamic/provider_test.go:271]: Reading this test standalone I can't figure out what it's testing and asserting. Perhaps a better name in TestCheckConfig or a quick comment.

File: pf/internal/configencoding/provider.go
Comment (t0yv0) [inline pf/internal/configencoding/provider.go:118]: Is attach overload needed?

File: pf/internal/plugin/provider_server.go
Comment (t0yv0) [inline pf/internal/plugin/provider_server.go:234]: I think this is reasonable. This entire file I understand will go away in favor of linking code directly from p/p at some point.

**Comment by t0yv0** [comment]
I'm working to get https://github.com/pulumi/pulumi/pull/15032 done next iteration. If this cannot wait I'm 👍 on shipping this change. I've reviewed step by step and it looks good!  The review is slow because I do not fully trust our coverage in this area, unfortunately, if we backfilled the coverage it would be so much easier to refactor. 

Along these lines, I'd need the following test coverage to be added before shipping.
 
1. this branch in config_encoding is not covered by unit tests according to gocov but needs to be:

```go
 if v.IsSecret() {
   isSecret = true
   v = v.SecretValue().Element
 }
```

2. similarly this branch is not covered:

```go
  case knownKey && v.IsNull():
                return resource.NewStringProperty(""), nil
```

3. This might be covered but I could not find it. We need to create an integration test for a PF provider that covers configuring the provider with nested values and nested value secrets specifically at integration level. Asserts: (1) the provider should receive the expected data in Configure and (2) the secret material should not leak into state. I looked some more but didn't find this integration level test for PF. This would be super helpful.

**Comment by t0yv0** [comment]
Nit: the new version of UnfoldProperties is more dense and less readable to me, but I think it does the same thing so not blocking on it. This bit is highly suspect legacy behavior and deserves to keep a comment I think, used to have a comment.

```
		if v.IsComputed() && v.V.(resource.Computed).Element.V == "" {
			v = resource.MakeComputed(enc.zeroValue(enc.fieldTypes[pk]))
		}
```

**Review by t0yv0 — APPROVED** [review]
LGTM if test coverage added as requested. Thank you!!

File: dynamic/provider_test.go
Comment (iwahbe) [reply dynamic/provider_test.go:271]: I'll add a comment.

File: pf/internal/configencoding/provider.go
Comment (iwahbe) [reply pf/internal/configencoding/provider.go:118]: It is not (and I removed it). The [`Attach` method](https://github.com/pulumi/pulumi-terraform-bridge/blob/63e2e72ebf1dba6cfa7cdec645ead9d21867009f/pf/internal/configencoding/provider.go#L30) on `ProviderWithContextAndConfig` is needed however. Go type casting is weird.

Explanation: https://go.dev/play/p/nfWG2seoG0s

**Comment by iwahbe** [comment]
> LGTM if test coverage added as requested. Thank you!!

Coverage is added in https://github.com/pulumi/pulumi-terraform-bridge/pull/2288. This PR is now rebased on top of https://github.com/pulumi/pulumi-terraform-bridge/pull/2288.

**Pull request merged**
